### ISSUE TITLE
Lookup program-to-cover in PATH unless full path is given

### ISSUE THREAD

**Issue by Mark-E-Hamilton** [SimonKagstrom/kcov#58]
We currently use the Python `coverage` packge  to check our python code, and parse the data files output by that tool to verify that all tested modules have the expected coverage. We'd like to do something similar with shell scripts, but since kcov doesn't generate a similarly formatted text file it seemed that the json file(s) might  be way to go. However, I tried a simple test to read the json fail which failed with this traceback.
````
 >>> obj = json.load(open("index.json"))
Traceback (most recent call last):
...
    raise ValueError("No JSON object could be decoded")
ValueError: No JSON object could be decoded
````
I then tried pasting the contents of the index.json file into  http://jsonlint.com and it would not validate there either. Am I doing this wrong, or do I just misunderstand what these files actually are? Is there  better way to parse them in python?

Also, is there a more appropriate forum for asking this sort of question instead of this issue tracker? 

Thanks.

Mark E. Hamilton

**Comment by SimonKagstrom** [comment]
The issue tracker is fine by me. Optionally, you could just send an email, but the traffic is so low anyway.

Anyway, I would go with the Cobertura output instead - it's in well defined XML. I use it in the automatic tests, with the help of a script in tests/tools/lookup-xml-node.py. The script prints how many hits a particular file/line has (or nocode if it isn't code), so I guess that should be pretty similar to what you're aiming at?

**Comment by Mark-E-Hamilton** [comment]
On 02/16/15 23:50, Simon Kagstrom wrote:
> The issue tracker is fine by me. Optionally, you could just send an
> email, but the traffic is so low anyway.
>
> Anyway, I would go with the Cobertura output instead - it's in well
> defined XML. I use it in the automatic tests, with the help of a script
> in tests/tools/lookup-xml-node.py. The script prints how many hits a
> particular file/line has (or nocode if it isn't code), so I guess that
> should be pretty similar to what you're aiming at?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SimonKagstrom/kcov/issues/58#issuecomment-74624706>.
>

Thanks. I don't know why I didn't see that before. ElementTree parses it 
nicely, so I can probably make it work for us.

I had another question, which I can make an issue if you'd like.

We have a test harness which exercises our shell scripts using Python's 
subprocess.Popen, basically executing them with some parameters, 
capturing the stdout and stderr, and comparing it to some expected 
results. I figured that I could just prepend  'kcov some_dir ' to the 
command to get coverage, but is looks like kcov doesn't search PATH for 
the passed file. (As I type this it tickles in the back of my mind that 
I might have asked this same question about shcov, but I don't recall 
clearly.)

I could probably do something like 'kcov some_dir $(type -p cmd) args 
...', but that seems really messy.

IAE, is there a reason not to scan PATH to find the file? If so, is it 
reasonable to consider an option to tell kcov to search PATH for the 
file, instead of looking for it in the filesystem, or even just 
searching PATH if the file has no directory information and isn't in the 
current directory?


-- 
----------------
Mark E. Hamilton
Engineering Sciences Center
Senior Member of Technical Staff
Sandia National Laboratories
505-844-7666

**Comment by SimonKagstrom** [comment]
Yes, that's a reasonable addition - I simply never thought about using it that way.

I'll take a look at falling back to $PATH/program if the program can't be found directly.

**Comment by SimonKagstrom** [comment]
And to add some advertisements: Kcov does Python code coverage collection as well, if you'd like to put all eggs in the same basket :-)

**Comment by Mark-E-Hamilton** [comment]
On 02/19/15 01:36, Simon Kagstrom wrote:
> And to add some advertisements: Kcov does Python code coverage
> collection as well, if you'd like to put all eggs in the same basket :-)

I've actually been thinking about that. I'd like to get the bash 
coverage going first, though. Thanks.

-- 
----------------
Mark E. Hamilton
Engineering Sciences Center
Senior Member of Technical Staff
Sandia National Laboratories
505-844-7666

**Comment by SimonKagstrom** [comment]
Commit 57973cc implements a fallback to look in $PATH if the program isn't explicitly given on the command line, so closing.

**Comment by Mark-E-Hamilton** [comment]
I ran into a minor issue with this change. We have a test which checks to ensure that non executable scripts aren't executed. However, when run with kcov using this change (ie, finding it in PATH) the script does get executed. In this case should it perhaps only run them if they are executable? Since kcov isn't a static analyzer, but actually executes the scripts passed, it seems logical to me.

Thanks.

**Comment by SimonKagstrom** [comment]
Yes, that sounds like reasonable behavior. I'll change it so that

1. If the script is passed directly on the command-line, it's always executed (i.e., the behavior before this change)

2. If the script is in the PATH and is executable, it will be executed, otherwise not.

**Comment by SimonKagstrom** [comment]
The executable problem has been fixed in e4f29d6.

**Comment by Mark-E-Hamilton** [comment]
Sorry, I hate to keep picking at this. I like the solution, but it does behave differently than I expected. For example, if I do this, kcov produces a different result than when I try to execute the test file:
{{{
% mkdir subdir
% echo -e '#!/bin/bash\necho Hi\n' > subdir/blather
% export PATH=${PWD}/subdir:$PATH
% type blather
blather is /sierra/dev/mhamilt/sntools_votd_5/subdir/blather
% blather
bash: /sierra/dev/mhamilt/sntools_votd_5/subdir/blather: Permission denied
% kcov /tmp/kcov_dir blather
Usage: kcov [OPTIONS] out-dir in-file [args...]
...
}}}
Basically, it says I didn't find it, not I found it but couldn't run it like bash does. It seems to me that a 'Permission denied' error would be more consistent with what bash does.

Thanks for all your work, BTW.

**Comment by SimonKagstrom** [comment]
Yes, a more logical error message would be good. I'll add that sometime in the future. I guess it's mostly a cosmetic issue though?

**Comment by Mark-E-Hamilton** [comment]
On 02/26/15 11:30, Simon Kagstrom wrote:
> Yes, a more logical error message would be good. I'll add that sometime
> in the future. I guess it's mostly a cosmetic issue though?

Yep. In fact, if I never see the message I don't care what it says. ;)

-- 
----------------
Mark E. Hamilton
Engineering Sciences Center
Senior Member of Technical Staff
Sandia National Laboratories
505-844-7666
